beam/core/src/main/java/cz/o2/proxima/beam/direct/io/OffsetRestrictionTracker.java

```diff
@@ -0,0 +1,164 @@
+import lombok.ToString;
+import org.apache.beam.sdk.transforms.splittabledofn.HasDefaultTracker;
+import org.apache.beam.sdk.transforms.splittabledofn.RestrictionTracker;
+import org.apache.beam.sdk.transforms.splittabledofn.SplitResult;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/**
+ * A {@link RestrictionTracker} for {@link Offset Offsets} read from {@link
+ * cz.o2.proxima.direct.commitlog.CommitLogReader}
+ */
+public class OffsetRestrictionTracker
+    extends RestrictionTracker<OffsetRestrictionTracker.StartEndOffset, Offset> {
+
+  @ToString
+  public static class StartEndOffset
+      implements Serializable, HasDefaultTracker<StartEndOffset, OffsetRestrictionTracker> {
+
+    private static final long serialVersionUID = 1L;
+
+    /**
+     * @return initial restriction for {@link
+     *     org.apache.beam.sdk.transforms.DoFn.GetInitialRestriction}.
+     */
+    public static StartEndOffset initialRestriction() {
+      return new StartEndOffset();
+    }
+
+    /** @return restriction that reads from given offset (exclusive) */
+    public static StartEndOffset startingFrom(Offset start) {
+      return new StartEndOffset(start, true);
+    }
+
+    @Getter @Nullable private final Offset startOffset;
+    @Getter private final boolean startInclusive;
+    @Getter @Nullable private volatile Offset endOffsetInclusive;
+    // weather to disallow adding claim past endOffset
+    @Getter private volatile boolean extensible;
+    private transient boolean finished = false;
+
+    private StartEndOffset(@Nullable Offset start, boolean inclusive) {
+      this.startOffset = start;
+      this.startInclusive = inclusive;
+      extensible = true;
+    }
+
+    private StartEndOffset() {
+      this(null, false);
+    }
+
+    private StartEndOffset(Offset startExclusive) {
+      this(Objects.requireNonNull(startExclusive), false);
+    }
+
+    public StartEndOffset(Offset startOffsetExclusive, Offset endOffsetInclusive) {
+      this.startOffset = Objects.requireNonNull(startOffsetExclusive);
+      this.startInclusive = false;
+      this.endOffsetInclusive = Objects.requireNonNull(endOffsetInclusive);
+      this.extensible = false;
+    }
+
+    boolean getAndCommitClaim(Offset offset) {
+      if (!extensible) {
+        if (finished) {
+          return false;
+        }
+        finished = offset.equals(endOffsetInclusive);
+        return true;
+      }
+      endOffsetInclusive = offset;
+      return true;
+    }
+
+    @Nullable
+    public StartEndOffset asPrimary() {
+      if (endOffsetInclusive == null) {
+        return null;
+      }
+      return new StartEndOffset(startOffset, endOffsetInclusive);
+    }
+
+    public StartEndOffset asResidual() {
+      return new StartEndOffset(endOffsetInclusive);
+    }
+
+    public boolean isInitial() {
+      return startOffset == endOffsetInclusive && endOffsetInclusive == null;
+    }
+
+    void terminate() {
+      this.extensible = false;
```

Comment:
is this flag really needed?

Reply:
yes

Reply:
We need that to enable `tryClaim` to return `false`.

Reply:
hm, that's currently handled by `terminate`, but this flag is needed to be able to reprocess already split restriction. That is probably not done (currently), but the contract allows for that.

Reply:
Can you please add javadoc explaining `terminate` and `extensible` flags? should we rename `extensible` to unbounded?